pkg/app/api/grpcapi/web_api.go

```diff
@@ -194,6 +197,83 @@ func (a *WebAPI) DeleteEnvironment(ctx context.Context, req *webservice.DeleteEn
 	if err := a.validateEnvBelongsToProject(ctx, req.EnvironmentId, claims.Role.ProjectId); err != nil {
 		return nil, err
 	}
+	// Check if no Piped has permission to the given environment.
+	pipeds, err := a.pipedStore.ListPipeds(ctx, datastore.ListOptions{
+		Filters: []datastore.ListFilter{
+			{
+				Field:    "EnvIds",
+				Operator: datastore.OperatorIn,
+				Value:    req.EnvironmentId,
+			},
+			{
+				Field:    "ProjectId",
+				Operator: datastore.OperatorEqual,
+				Value:    claims.Role.ProjectId,
+			},
+		},
+	})
+	if err != nil {
+		a.logger.Error("failed to fetch Pipeds linked to the given environment",
+			zap.String("env-id", req.EnvironmentId),
+			zap.Error(err),
+		)
+		return nil, status.Error(codes.Internal, "Failed to validate the deletion operation")
+	}
+	if len(pipeds) > 0 {
+		var pipedNames strings.Builder
+		for _, p := range pipeds {
+			pipedNames.WriteString(fmt.Sprintf("%q", p.Name))
+		}
+		return nil, status.Errorf(
+			codes.FailedPrecondition,
+			"Found Pipeds linked the environment to be deleted. Please remove this environment from all Pipeds (%s) on the Piped settings page",
+			pipedNames,
+		)
+	}
+
+	// Delete all applications that belongs to the given env.
+	apps, _, err := a.applicationStore.ListApplications(ctx, datastore.ListOptions{
+		Filters: []datastore.ListFilter{
+			{
+				Field:    "EnvId",
+				Operator: datastore.OperatorEqual,
+				Value:    req.EnvironmentId,
+			},
+			{
+				Field:    "ProjectId",
+				Operator: datastore.OperatorEqual,
+				Value:    claims.Role.ProjectId,
+			},
+		},
+	})
+	if err != nil {
+		a.logger.Error("failed to fetch applications that belongs to the given environment",
+			zap.String("env-id", req.EnvironmentId),
+			zap.Error(err),
+		)
+		return nil, status.Error(codes.Internal, "Failed to fetch applications that belongs to the given environment")
+	}
+	for _, app := range apps {
+		if app.ProjectId != claims.Role.ProjectId {
+			return nil, status.Errorf(codes.PermissionDenied, "application %q doesn't belong to the project you logged in", app.Name)
+		}
```

Comment:
will this check is safe to be removed 🤔

Reply:
I feel you. Logically all applications should be within the scope of the project where the environment exists. But it's a deletion operation. I thought it's safer to expect that data inconsistency could happen for some reason.

Reply:
In that case, maybe just skip this app and continue deleting the others from this apps array could be enough, I guess. At least, we can remove all possible to be removed apps and the things that remained just some apps that really not belong to the claimed project 😄

Reply:
yeah, we can continue removing the remaining apps after that as you mentioned. But this data inconsistency is an extremely significant, fatal issue. I'd say we must immediately stop the operation and give back the permission error, but what would you say?

Reply:
Since this is soft delete, I think the current check is safe enough for this action. Currently, we first filter by the claimed project, and secondary compare this queried app's projectID here. That was really strict already 😂

Reply:
>instead of returning just skip this app and continue deleting the others

Fine, I pick up what you second suggested 👍

Reply:
thank you 🙏